src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,806 @@
+                        {
+                            if (asname == null)
+                            {
+                                statement = String.Format("import {0}", memberName);
+                            }
+                            else
+                            {
+                                statement = String.Format("import {0} as {1}", memberName, asname);
+                            }
+                        }
+                        else
+                        {
+                            if (memberName != "*" && asname != null)
+                            {
+                                statement = String.Format("from {0} import {1} as {2}", module, memberName, asname);
+                            }
+                            else
+                            {
+                                statement = String.Format("from {0} import *", module);
+                            }
+                        }
+
+                        BadStatements.TryGetValue(statement, out previousTries);
+
+                        if (previousTries > 3)
+                        {
+                            continue;
+                        }
+
+                        Scope.Exec(statement);
+
+                        if (memberName == "*")
+                        {
+                            continue;
+                        }
+
+                        string typeName = module == null ? memberName : String.Format("{0}.{1}", module, memberName);
+                        var type = Type.GetType(typeName);
+                        ImportedTypes.Add(name, type);
+                    }
+                    catch (Exception)
+                    {
+                        Log(String.Format("Failed to load module: {0}, with statement: {1}", memberName, statement));
+                        // Log(e.ToString());
+                        BadStatements[statement] = previousTries + 1;
+                    }
+                }
+            }
+            finally
+            {
+                PythonEngine.ReleaseLock(gs);
+            }
+        }
+
+        /// <summary>
+        /// Traverse the given source code and define variable types based on
+        /// the current scope
+        /// </summary>
+        /// <param name="code">The source code to look through</param>
+        private void UpdateVariableTypes(string code)
+        {
+            VariableTypes.Clear();
+            VariableTypes = FindAllVariableAssignments(code);
+        }
+
+        /// <summary>
+        /// Attempts to find all variable assignments in the code. Has basic variable unpacking support.
+        /// We don't need to check the line indices because regex matches are ordered as per the code.
+        /// </summary>
+        /// <param name="code">The code to search</param>
+        /// <returns>A dictionary of variable name and type pairs</returns>
+        private Dictionary<string, Type> FindAllVariableAssignments(string code)
```

Comment:
not something for today I guess - but it would be really nice to replace all these complex regex matches throughout these classes with https://docs.python.org/3/library/ast.html#ast.parse